File: server/services/identity-service/SafeRide.Identity/SafeRide.Identity.Application/Common/Exceptions/AppException.cs
namespace SafeRide.Identity.Application.Common.Exceptions;

public class AppException(string code, string message, int statusCode = 400) : Exception(message)
{
    public string Code { get; } = code;
    public int StatusCode { get; } = statusCode;
}

public sealed class NotFoundException(string message, string code = "Resource.NotFound")
    : AppException(code, message, 404);

public sealed class ConflictException(string message, string code = "Resource.Conflict")
    : AppException(code, message, 409);

public sealed class ForbiddenException(string message, string code = "Auth.Forbidden")
    : AppException(code, message, 403);

File: server/services/route-service/route/src/main/java/com/saferide/route/exception/AppException.java
package com.saferide.route.exception;

import lombok.Getter;

@Getter
public class AppException extends RuntimeException {
    private final String code;
    private final int statusCode;

    public AppException(String code, String message, int statusCode) {
        super(message);
        this.code = code;
        this.statusCode = statusCode;
    }

    public static class ConflictException extends AppException {
        public ConflictException(String message) {
            super("Resource.Conflict", message, 409);
        }
    }

    public static class ForbiddenException extends AppException {
        public ForbiddenException(String message) {
            super("Auth.Forbidden", message, 403);
        }
    }

    public static class NotFoundException extends AppException {
        public NotFoundException(String message) {
            super("Resource.NotFound", message, 404);
        }
    }
}
